### PR TITLE
Add support for random reading of bgzipped FASTA files.

### PR DESCRIPTION
**Pull request by alumi**
#### Summary
This PR adds support for random reading of bgzip-compressed FASTA files.
`cljam.io.sequence/read-sequence` now works on `fa.gz` files if `fa.gz.fai` and `fa.gz.gzi` exist.
`.gzi` is a file format for storing BGZF indices as a list of pairs (compressed offset, uncompressed offset).
Note that this implementation does not support indexing FASTA without `.gzi` file.

#### Tests

- `lein check` ✅ 
- `lein test :all` ✅ 
- `lein cloverage` ✅

### CODE REVIEW

File: src/cljam/io/util/bgzf/gzi.clj [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  "Returns a virtual file offset for a given uncompressed offset."
+  ^long [gzi ^long uncompressed-offset]
+  (let [[u c] (first (rsubseq gzi <= uncompressed-offset))
+        diff (- uncompressed-offset (long u))]
+    (assert (<= 0 diff 0xffff))
+    (bit-or (bit-shift-left (long c) 16)
+            (bit-and diff 0xffff))))
+
+(defn comp->uncomp
+  "Returns a uncompressed offset for a given virtual file offset"
```
Comment (athos) [inline src/cljam/io/util/bgzf/gzi.clj]:
```suggestion
  "Returns an uncompressed offset for a given virtual file offset."
```

File: src/cljam/io/fasta/reader.clj [hunk 2]
```diff
@@ -1,25 +1,65 @@
 (ns cljam.io.fasta.reader
-  (:refer-clojure :exclude [read])
+  (:refer-clojure :exclude [read read-line])
   (:require [cljam.util :refer [graph?]]
             [cljam.io.fasta.util :refer [header-line? parse-header-line]]
-            [cljam.io.fasta-index.core :as fasta-index])
-  (:import [java.io RandomAccessFile InputStream]
+            [cljam.io.fasta-index.core :as fasta-index]
+            [cljam.io.util.bgzf.gzi :as gzi]
+            [cljam.io.util.bgzf :as bgzf])
```
Comment (r6eve) [inline src/cljam/io/fasta/reader.clj]: `cljam.io.util.bgzf` is never used.

File: src/cljam/io/util/bgzf/gzi.clj (hunk 1 above)
Comment (alumi) [reply src/cljam/io/util/bgzf/gzi.clj]: Thanks! Fixed in c08fa83.

File: src/cljam/io/fasta/reader.clj (hunk 2 above)
Comment (alumi) [reply src/cljam/io/fasta/reader.clj]: Good catch! Thanks 🙏

File: src/cljam/io/util/bgzf/gzi.clj [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    (bit-or (bit-shift-left (long c) 16)
+            (bit-and diff 0xffff))))
+
+(defn comp->uncomp
+  "Returns an uncompressed offset for a given virtual file offset"
+  ^long [gzi ^long compressed-offset]
+  (let [off (unsigned-bit-shift-right compressed-offset 16)
+        [uncompressed] (->> gzi
+                            rseq
+                            (filter (fn [[u c]] (= (long c) off)))
```
Comment (r6eve) [inline src/cljam/io/util/bgzf/gzi.clj]: `u` is never used. https://guide.clojure.style/#underscore-for-unused-bindings

**Comment by alumi** [comment]
@r6eve @yito88 @athos Thank you for reviewing!

**Pull request merged**